legacy/reqcounter/reqcounter_test.go

```diff
@@ -0,0 +1,191 @@
+			},
+			{
+				Mutex:    sync.Mutex{},
+				Requests: requests[1],
+				Since:    time.Now(),
+				Interval: time.Second,
+			},
+			{
+				Mutex:    sync.Mutex{},
+				Requests: requests[2],
+				Since:    time.Now(),
+				Interval: time.Second,
+			}},
+	}
+	// Ensure that modifying counters can only occur when counting is enabled.
+	rc.EnableCounting = false
+	rc.Increment()
+	for i, counter := range rc.Monitor.Counters {
+		if counter.Requests != requests[i] {
+			fmt.Println("Counters were modified while counting was disabled")
+			t.Fail()
+		}
+	}
+	// Ensure the Increment function actually increments each counter's requests field.
+	rc.EnableCounting = true
+	rc.Increment()
+	for i, counter := range rc.Monitor.Counters {
+		if counter.Requests != requests[i]+1 {
+			fmt.Printf("Counters were not incremented correctly. Expected: %d Got: %d\n", requests[i]+1, counter.Requests)
+			t.Fail()
+		}
+	}
+}
+
+func TestFlush(t *testing.T) {
+	// Create a local invocation of time.
+	currentTime := time.Now()
+	counter := &rc.RequestCounter{
+		Mutex:    sync.Mutex{},
+		Requests: 100,
+		Since:    currentTime,
+		Interval: time.Second,
+	}
+	// Mock the logrus.Debug function.
+	debugCalls := 0
+	rc.Debug = func(args ...interface{}) {
+		debugCalls++
+	}
+	// Flush the request counter.
+	counter.Flush()
+	// Ensure logrus.Debug was called.
+	if debugCalls != 1 {
+		fmt.Println("Flush failed to trigger a debug statement.")
+		t.Fail()
+	}
+	// Ensure the counter is reset, where time advances and the requests are zeroed.
+	counter.Flush()
+	if counter.Requests != 0 || !currentTime.Before(counter.Since) {
+		fmt.Println("Calling Flush did not reset the counter.")
+		t.Fail()
+	}
+}
+
+func TestLog(t *testing.T) {
+	// Create two counters with different logging intervals. During testing, only one
+	// of the counters should fire logrus.Debug statements. To determine which counter
+	// was logged, we can use the Requests number to identify them.
+	unexpected := uint64(80808)
+	expected := uint64(33333)
+	netMonitor := &rc.NetworkMonitor{
+		Counters: []*rc.RequestCounter{
+			{
+				Mutex:    sync.Mutex{},
+				Requests: unexpected,
+				Since:    time.Time{},
+				Interval: time.Hour,
+			},
+			{
+				Mutex:    sync.Mutex{},
+				Requests: expected,
+				Since:    time.Time{},
+				Interval: time.Second,
+			},
+		},
+	}
+	// Mock the logrus.Debug function, ensuring the correct counter is logged.
+	sawExpectedLog := false
+	rc.Debug = func(args ...interface{}) {
+		// Convert to string for easier parsing.
+		s := fmt.Sprint(args)
+
+		if strings.Contains(s, fmt.Sprint(unexpected)) {
+			fmt.Println("The counter with 'Interval: 1hr' was prematurely logged after waiting only 3 seconds.")
+			t.Fail()
+		} else if strings.Contains(s, fmt.Sprint(expected)) {
+			sawExpectedLog = true
+		}
+	}
+	// Begin logging each counter.
+	netMonitor.Log()
+	// Allow enough time to ensure ONLY the expected counter can log itself.
+	time.Sleep(time.Second * 3)
```

Comment:
Currently the RequestCounters call time.Sleep() to pause between each interval. Instead you can give it either a real sleep function (time.Sleep()) or a fake sleep function (make one up). For production you can pass in the real sleep function, but for testing you can pass in a fake sleep function that returns instantaneously. Then we don't have to sleep 3 seconds for this single unit test (quite expensive).

In general any sort of reliance on actual real-world events (in this case, sleeping for a number of seconds) in testing code is an anti-pattern.

Reply:
> In general any sort of reliance on actual real-world events (in this case, sleeping for a number of seconds) in testing code is an anti-pattern.

I should clarify that this is the case for unit tests. For larger tests (e.g., integration, e2e) your scope increasingly includes the real world, so the above statement loses significance.

Reply:
I shortened the time to sleep, which should still be more than enough time to reliably log all counters. If this is still a concern, please bring this to my attention. Resolving.